### PR TITLE
feat: support recovering from a volume snapshot backup object

### LINKED ISSUE #2964: Properly manage volume snapshot method when recovering from a backup object
Currently, when recovering from a backup object (which will be deprecated at some point) we assume that the backup is in an object store. Introduce a check to ensure that recovery can also cater for volume snapshots. Blocked by #2963

### PR DESCRIPTION
**Pull request by mnencia**
Closes #2964

### CODE REVIEW

**Comment by mnencia** [comment]
E2E tests running on EDB fork https://github.com/EnterpriseDB/cloudnative-pg/actions/runs/6656938711

**Review by armru — APPROVED** [review]

File: pkg/reconciler/persistentvolumeclaim/storagesource.go [hunk 1]
```diff
@@ -131,33 +133,35 @@ func getCandidateSourceFromBackupList(ctx context.Context, backupList apiv1.Back
...
-				Kind:     "VolumeSnapshot",
-				Name:     GetName(backup.Name, utils.PVCRolePgData),
-			},
+		return getCandidateSourceFromBackup(backup)
+	}
+
+	return nil
+}
+
+func getCandidateSourceFromBackup(backup *apiv1.Backup) *StorageSource {
```
Comment (armru) [inline pkg/reconciler/persistentvolumeclaim/storagesource.go:142]: in future, we should also return an error if there is no match for the element.type

Comment (mnencia) [reply pkg/reconciler/persistentvolumeclaim/storagesource.go:142]: Yes. Especially when we add the tablespaces, which have one more degree of freedom (the tablespace name)

**Review by leonardoce — APPROVED** [review]

**Comment by gbartolini** [comment]
I am not a big fan of this approach, but I see that others are OK with it, so I approve it.

**Pull request merged**